### ISSUE TITLE
Consider support for mocking websockets

### ISSUE THREAD

**Issue by blueridanus** [LukeMathWalker/wiremock-rs#113]
Would this addition be welcome? I could look into writing a PR for this.

**Comment by LukeMathWalker** [comment]
I'd like to understand the design before going through a PR!
I'm definitely interested in the feature 😄

**Comment by LukeMathWalker** [comment]
What does `ws::EchoResponder` look like in terms of implementation? What trait is it implementing?

**Comment by blueridanus** [comment]
I figured something similar, but websocket specific (e.g. `WsRespond`). I think the more generic approach above could be even better, as long some attention is also paid to the possibility of `Transfer-Encoding: chunked` (which that trait name seems to suggest).

**Comment by xd009642** [comment]
Yeah, I think for other things such as websockets or grpc you would need to bring in a library like tungstenite and tonic for the transport stuff that's not as visible to user (i.e. websocket handshaking etc). So they'd have to be optional dependencies (no need to add bloat for people).

I guess that means adding some sort of transport style middleware to handle serializing/deserializing to things like websocket TCP frames and also doing any handshaking etc that has to be done. Maybe it would look a bit like:

```rust
use wiremock::ws::*;

#[async_trait]
impl StreamingRespond for Streamer {
   // left as an exercise to the reader
}

Mock::given(ws::Upgrade)  // new kind of matcher
    .and(path("/hello_ws"))
    // this could simply echo ws messages back to client. could also have another case which takes a closure
    .transport(transport::Websocket)
    .streaming_respond_with(Streamer::default())
    .mount(&mock_server)
    .await;
```

**Comment by blueridanus** [comment]
Another advantage of forgoing the generic approach would be handlers/implementors getting typed messages instead of raw bytes, although this could potentially be addressed by making the `StreamingRespond` trait even more generic.

**Comment by LukeMathWalker** [comment]
I lean towards having the interface being specific to websockets—we can always generalize later if/when we decide to support something like gRPC (which hasn't been the case for the past ~3 years).
Especially if a special-cased interface allows us to provide better ergonomics, which is the paramount focus in `wiremock`—writing tests is annoying enough.

**Comment by dpezely** [comment]
I can help if you need early adopters to test and contribute based upon some familiar with WebSockets as well as using `wiremock`.

(Currently working on a project that could use this, and we're already using `wiremock` thus would like to keep it all in same the family.  Thank you for doing this!)

**Comment by xd009642** [comment]
For those who come upon this issue I've made https://github.com/xd009642/wiremocket which was an experiment in how a PR could look but is currently it's own project. Not opposed to merging them if this pans out but I figured it needed more incubation

**Comment by georglauterbach** [comment]
> For those who come upon this issue I've made https://github.com/xd009642/wiremocket which was an experiment in how a PR could look but is currently it's own project. Not opposed to merging them if this pans out but I figured it needed more incubation

I am also interested in websocket support, and I will be trying `wiremocket` and report back here.

**Comment by xd009642** [comment]
Instead of reporting back here, feel free to use the wiremock issues. It'll save Luca some needless notifications 😅
